### PR TITLE
Support some of aggregation functions for interval types

### PR DESCRIPTION
**Pull request by harunurhan**
#4417

I re-used existing aggregation implementations for all of them. 

Functions:
- variance
- stddev
- var_pop
- stddev_pop
- avg
- geometric_mean

Please note that I haven't done manual testing for all of them, other than running some queries like `select avg(INTERVAL '123-3' YEAR TO MONTH);`

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/AverageAggregations.java [hunk 1]
```diff
@@ -32,14 +32,28 @@
...
     {
         state.setLong(state.getLong() + 1);
         state.setDouble(state.getDouble() + value);
     }
 
     @InputFunction
-    public static void input(LongAndDoubleState state, @SqlType(StandardTypes.DOUBLE) double value)
+    public static void doubleInput(LongAndDoubleState state, @SqlType(StandardTypes.DOUBLE) double value)
+    {
+        state.setLong(state.getLong() + 1);
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/AverageAggregations.java]: Use `Math.incrementExact` (https://docs.oracle.com/javase/8/docs/api/java/lang/Math.html#incrementExact-int-) to cover potential overflow.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/GeometricMeanAggregations.java [hunk 2]
```diff
@@ -26,14 +26,28 @@
...
     {
         state.setLong(state.getLong() + 1);
         state.setDouble(state.getDouble() + Math.log(value));
     }
 
     @InputFunction
-    public static void input(LongAndDoubleState state, @SqlType(StandardTypes.DOUBLE) double value)
+    public static void doubleInputInput(LongAndDoubleState state, @SqlType(StandardTypes.DOUBLE) double value)
+    {
+        state.setLong(state.getLong() + 1);
```
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/GeometricMeanAggregations.java]: ditto

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestIntervalDayToSecondVarianceAggregation.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        Variance variance = new Variance();
+        return variance.evaluate(values);
+    }
+
+    @Override
+    protected String getFunctionName()
+    {
+        return "variance";
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestIntervalDayToSecondVarianceAggregation.java]: Similarly to `stddev`, shouldn't we name it `var_samp` here, just to be consistent? Especially that you provided `var_pop` below.

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestIntervalYearToMonthAverageAggregation.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        for (int i = start; i < start + length; i++) {
+            INTERVAL_YEAR_MONTH.writeLong(blockBuilder, i);
+        }
+        return new Block[] {blockBuilder.build()};
+    }
+
+    @Override
+    public Number getExpectedValue(int start, int length)
+    {
+        if (length == 0) {
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestIntervalYearToMonthAverageAggregation.java]: Isn't `< 1` safer?

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestIntervalYearToMonthStdDevPopAggregation.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        for (int i = start; i < start + length; i++) {
+            INTERVAL_YEAR_MONTH.writeLong(blockBuilder, i);
+        }
+        return new Block[] {blockBuilder.build()};
+    }
+
+    @Override
+    public Number getExpectedValue(int start, int length)
+    {
+        if (length == 0) {
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestIntervalYearToMonthStdDevPopAggregation.java]: ditto `< 1`

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestIntervalYearToMonthVarianceAggregation.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        Variance variance = new Variance();
+        return variance.evaluate(values);
+    }
+
+    @Override
+    protected String getFunctionName()
+    {
+        return "variance";
```
Comment (maciejgrzybek) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestIntervalYearToMonthVarianceAggregation.java]: ditto `var_samp`

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestIntervalDayToSecondVarianceAggregation.java (hunk 3 above)
Comment (harunurhan) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestIntervalDayToSecondVarianceAggregation.java]: @maciejgrzybek `variance` is alias of `var_samp`, and `variance` is used in other tests such as [TestLongVarianceAggregation.java#L59](https://github.com/facebook/presto/blob/master/presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestLongVarianceAggregation.java#L59)

File: presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestIntervalYearToMonthAverageAggregation.java (hunk 4 above)
Comment (harunurhan) [inline presto-main/src/test/java/com/facebook/presto/operator/aggregation/TestIntervalYearToMonthAverageAggregation.java]: @maciejgrzybek might be, but I simply copied implementation from other tests. So may be another issue should be create for this.

If you create issue for improving these tests,
Also: we should find a way to share test implementation of the same function among tests for different types such as `TestLongVarianceAggregation`, `TestDoubleVarianceAggregation`, `TestIntervalYearToMonthAverageAggregation ` etc. `getExpectedValue` and `getFunctionName` methods are almost always the same.

**Comment by martint** [comment]
@cberner can you review?

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/AverageAggregations.java (hunk 1 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/AverageAggregations.java]: We would actually want to make our own version that threw `PrestoException` with an error code. We probably want to benchmark to make sure that doesn't hurt performance though, and then change it in all the aggregations. Can you file a ticket for that?

**Comment by cberner** [comment]
We're discussing options for implementing these. The signatures are quite strange at the moment, since they take `INTERVAL` types, but return `DOUBLE`. It seems like they should return `INTERVAL`, except for `variance` which should return `INTERVAL^2`

**Comment by cberner** [comment]
@electrum or @martint could you look at the signature of these and decide if we want to merge them? I can do the review, but the signatures (as currently implemented) don't make sense, imo

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/AverageAggregations.java (hunk 1 above)
Comment (maciejgrzybek) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/AverageAggregations.java]: Sorry for the delay. It's here: https://github.com/prestodb/presto/issues/5727

**Comment by martint** [comment]
Closing this one. We'll get merge https://github.com/prestodb/presto/pull/7026, instead.

**Pull request closed without merging**